### PR TITLE
Vault Dependency Upgrades [VAULT-871]

### PR DESCRIPTION
**Pull request by HridoyRoy**
This set of dependency upgrades fixes the known CVEs in our libraries: https://hashicorp.atlassian.net/browse/VAULT-871 . The only dependency not upgraded is `github.com/opencontainers/runc` which I believe is only used in our docker tests. The reason I couldn't upgrade runc is in a comment below.

This change is large, but should not be too risky, as the major changes are mostly changes to our testing. There seem to be a few rabbitmq changes as well, that should be tested manually, and vault on windows should also be tested manually due to the winio upgrade.

### CODE REVIEW

**Comment by HridoyRoy** [comment]
So far this upgrades the following top level dependencies, and all nested dependencies:

`github.com/hashicorp/vault-plugin-auth-cf
github.com/hashicorp/vault-plugin-auth-centrify
github.com/hashicorp/vault-plugin-auth-kerberos
github.com/hashicorp/vault-plugin-database-elasticsearch
github.com/hashicorp/vault-plugin-secrets-alicloud
github.com/hashicorp/vault-plugin-secrets-kv
github.com/hashicorp/vault-plugin-auth-azure
github.com/hashicorp/vault-plugin-auth-alicloud
github.com/hashicorp/vault-plugin-database-mongodbatlas
github.com/hashicorp/vault-plugin-secrets-ad
github.com/hashicorp/vault-plugin-secrets-openldap
github.com/hashicorp/vault-plugin-secrets-gcp
github.com/hashicorp/vault-plugin-auth-jwt
github.com/hashicorp/vault-plugin-secrets-mongodbatlas`

However, this manually fixes the version for:

`github.com/coreos/etcd
go.etcd.io/bbolt
go.etcd.io/etcd
google.golang.org/grpc`

due to: https://github.com/etcd-io/etcd/issues/12569 .

**Comment by HridoyRoy** [comment]
github.com/opencontainers/runc can’t be upgraded right now due to the following error:

`go: found github.com/opencontainers/runc/libcontainer/nsenter in github.com/opencontainers/runc v1.0.0-rc9
package github.com/opencontainers/runc/libcontainer/nsenter: C source files not allowed when not using cgo or SWIG: cloned_binary.c nsexec.c`

File: go.mod
Comment (ncabatoff) [inline go.mod:224]: This seems misleading.  We're replacing our existing dependency on 1.29.1 with 1.35.0, but then in the replace directive above we're forcing it to 1.27.1?  Can you see if it still works if you change both here and the replace directive to be 1.29.1?

File: go.mod
Comment (ncabatoff) [inline go.mod:209]: Do we maybe skip etcd for now?  Waiting might allow us to avoid the whole replace nonsense which feels like a bit of a minefield.

File: go.mod
Comment (HridoyRoy) [reply go.mod:224]: Yeah, I definitely see that. I think setting that to 1.29.1 should work.

File: go.mod
Comment (HridoyRoy) [reply go.mod:209]: I think the replace directive mitigates https://github.com/etcd-io/etcd/security/advisories/GHSA-chh6-ppwq-jh92 , which is a moderate vuln but is marked as high on the spreadsheet. I'll ask if we can skip this one, but I do agree that it's kind of a minefield.

File: vault/testing.go
Comment (ncabatoff) [inline vault/testing.go:395]: I'm not seeing why these are needed?

Comment (HridoyRoy) [reply vault/testing.go:395]: Ah, so github.com/mitchellh/go-testing-interface v1.14.1 has updated the T interface to include a Parallel() method. This method does not exist in the TB interface in the native testing library, and so methods that were originally using github.com/mitchellh/go-testing-interface v1.14.1's T interface and being called by functions that needed to use the TB interface because of benchmarking tests no longer worked. 

The solution I thought of was just to wrap the methods that require the github.com/mitchellh/go-testing-interface v1.14.1 T interface with a "Benchmarking" function that takes in a TB interface parameter and casts it to a github.com/mitchellh/go-testing-interface v1.14.1 T interface. In the case where a native Golang testing T struct is used, everything will work fine. In the case where a B struct is used, B will get cast to TB, which can be cast to T because T has all the methods in TB, but not vice versa.

Comment (ncabatoff) [reply vault/testing.go:395]: Ok, thanks for explaining.  I'm not thrilled with this solution, but it'll do and I don't have a better one off the top of my head.  I mean we could PR a TB type into go-testing-interface and use that in all the test helper signatures, but I don't want to block this PR just for that.

File: go.mod
Comment (ncabatoff) [reply go.mod:209]: I don't see the endorsement (I don't think the person who suggested it is part of the etcd dev team), but either way, my main concern is, might this prevent us from upgrading boltdb or grpc if we need to for a more serious reason?  These are way more essential components of Vault than etcd.

File: builtin/logical/transit/path_encrypt.go
Comment (ncabatoff) [inline builtin/logical/transit/path_encrypt.go:162]: How was this necessary for bumping dependencies?

Comment (HridoyRoy) [reply builtin/logical/transit/path_encrypt.go:162]: TestTransit_decodeBatchRequestItems has the following check: 

`			expectedErr := mapstructure.Decode(tt.src, &expectedDest)

			gotErr := decodeBatchRequestItems(tt.src, &tt.dest)
			gotDest := tt.dest

			if !reflect.DeepEqual(expectedErr, gotErr) {
				t.Errorf("decodeBatchRequestItems unexpected error value, want: '%v', got: '%v'", expectedErr, gotErr)
			}`

which fails without these changes like so: https://app.circleci.com/pipelines/github/hashicorp/vault/15230/workflows/0b4e1411-9665-4b36-9991-e6cac823a4ac/jobs/149962

Since the comment above decodeBatchRequestItems says: 

`
// It aims to behave as closely possible to the original mapstructure.Decode 
`

I figured I should make the two error messages match instead of changing the test.

Comment (ncabatoff) [reply builtin/logical/transit/path_encrypt.go:162]: Fair enough.

**Review by ncabatoff — APPROVED** [review]
One outstanding comment I don't feel too strongly about.  Probably best to have another approver before merging though.

File: go.mod
Comment (HridoyRoy) [reply go.mod:225]: I tried to change this to 1.29.1, but doing a go mod vendor after changing this value changes it back to 1.35.0.

Comment (ncabatoff) [reply go.mod:225]: Ok, don't worry about it then.

**Pull request merged**
